deployment/src/main/java/io/quarkiverse/jackson/jq/deployment/JacksonJqProcessor.java

```diff
@@ -1,59 +1,146 @@
-    }
+        IndexView indexView = combinedIndex.getIndex();
+        List<String> excludes = config.functions.excludes.orElseGet(Collections::emptyList);
+
+        RuntimeValue<Scope> root = recorder.createScope();
+        RuntimeValue<Scope> local = recorder.createScope(root);
+
+        ClassLoader cl = BuiltinFunctionLoader.class.getClassLoader();
+
+        // load build int functions
+        loadFunctionsFromConfig(cl, config.functions.version).forEach(e -> {
+            if (!excludes.contains(e.name)) {
+                recorder.addFunction(root, e.name, e.args, e.body, config.functions.version.toString());
+            }
+        });
+        lookupFunctions(indexView, config, context, BuiltinFunction.class).forEach(f -> {
+            if (!excludes.contains(StringUtils.substringBefore(f.getName(), '/'))) {
+                recorder.addFunction(local, f.getName(), f.getFunction());
+            }
+        });
+        lookupFunctions(indexView, config, context, net.thisptr.jackson.jq.internal.BuiltinFunction.class).forEach(f -> {
+            if (!excludes.contains(StringUtils.substringBefore(f.getName(), '/'))) {
+                recorder.addFunction(local, f.getName(), f.getFunction());
+            }
+        });
+
+        // load custom function
+        lookupFunctions(indexView, config, context, JqFunction.class).forEach(f -> {
+            if (!excludes.contains(StringUtils.substringBefore(f.getName(), '/'))) {
+                recorder.addFunction(local, f.getName(), f.getFunction());
+            }
+        });
+
+        // load other function
+        functions.forEach(f -> {
+            if (!excludes.contains(StringUtils.substringBefore(f.getName(), '/'))) {
+                recorder.addFunction(local, f.getName(), f.getFunction());
+            }
+        });
 
-    @BuildStep
-    @Record(ExecutionTime.STATIC_INIT)
-    SyntheticBeanBuildItem quarkusScopeBean(JacksonJqScopeRecorder recorder,
-            RecorderContext context) throws NoSuchMethodException {
-        // preload everything and use it as a parent scope
-        final Map<String, Function> functions = BuiltinFunctionLoader.getInstance()
-                .loadFunctionsFromServiceLoader(BuiltinFunctionLoader.class.getClassLoader(), Versions.JQ_1_6);
         return SyntheticBeanBuildItem
                 .configure(Scope.class)
                 .scope(Singleton.class)
-                .runtimeValue(recorder.createScope(functions))
+                .runtimeValue(local)
                 .done();
     }
+
+    @BuildStep
+    UnremovableBeanBuildItem unremovableBeans() {
+        return UnremovableBeanBuildItem.beanTypes(Scope.class);
+    }
+
+    @BuildStep
+    AdditionalApplicationArchiveMarkerBuildItem applicationArchiveMarkers() {
+        return new AdditionalApplicationArchiveMarkerBuildItem("net/thisptr/jackson/jq/Scope.class");
+    }
+
+    static List<JacksonJqFunctionBuildItem> lookupFunctions(
```

Comment:
discovery is based on jandex here so we can skip doing service loader dicovery as in any case, the original code would skip any service that has no annotation.

We may add jaandex support tot he underlying library